apps/librelingo_json_export/librelingo_json_export/export.py

```diff
@@ -46,16 +47,18 @@ def _export_skill(export_path: str, skill: Skill, course: Course, settings=None)
             json.dump(skill_data, f, ensure_ascii=False, indent=2)
 
     if skill.introduction:
-        Path(Path(export_path) / "introduction").mkdir(parents=True, exist_ok=True)
+        (Path(export_path) / "introduction").mkdir(parents=True, exist_ok=True)
+        output_path = (
+            Path(os.devnull)
```

Comment:
it seems like you are only doing this logic for the introduction files `.md` files but not for the json files in the output :thinking: which makes me think that for some reason the test is still wrong, because it doesn't fail. it should fail because I think according to the code, the JSON files are still created, so probably in the test JSON files would not be created (or perhaps, no files would be created) even if you don't pass the dry run flag.

Can you please try running the test, but change 

```python
invoke([*inputs, "--dry-run"])
```

to 

```python
invoke([*inputs])
```

This should definitely fail. If it doesn't then it means either the test setup is completely wrong, or the assertion is completely wrong.

So please also take a look at the assertion:

```python
assert os.listdir(".") == ["tmp"]
```

Perhaps `os.listdir(".")` will always return `["tmp"]` here no matter whether the files are generated or not.  Maybe it needs to be 

```python
assert os.listdir("./tmp") == [
  # something something
]
```

Reply:
I already did this experiment before. The assertion was failing, because `os.listdir(".")` was returning something like `["tmp", "FakeValue(id=123)"]`.

Note that in case of `json`'s the function `json.dumps` is used (cf. [here](https://github.com/LibreLingo/LibreLingo/blob/61fc29c9c75b4240adbfba66b8a385c4f1420de3/apps/librelingo_json_export/librelingo_json_export/export.py#L44) and [here](https://github.com/LibreLingo/LibreLingo/blob/61fc29c9c75b4240adbfba66b8a385c4f1420de3/apps/librelingo_json_export/librelingo_json_export/export.py#L76)).

I will paste the logs of the _failed run_ in a moment.

Reply:
I am thinking, though, that an entire file structure should be tested here, no? Perhaps more like an `os.walk()`, so as to make sure that actually no new files are created.

Or even do sth like 

```python
files_before_dry_run = set(os.walk())
# ... etc ...

assert set(os.walk()) == files_before_dry_run

```

Reply:
In the newest change, I am comparing the result of `tuple(os.walk("."))`. I think this will also help in issue #2479, since "tmp" is not hard coded.

Reply:
Gotcha, I'm just wondering if the test is working overall, does it fail if the json files are accidentally saved? :thinking: and it would be nice to refactor that part the same way for consistency, or even extract this logic somehow to make the code look neater. I can jump on a call with you to talk about it if you want: https://app.harmonizely.com/harmonizely-uoicqk

Reply:
You are right, the test passes if some json files are accidentally created.